### PR TITLE
All changelisteners are now delayed until next looper event

### PR DESCRIPTION
**Pull request by cmelchior**
Prerequisite for https://github.com/realm/realm-java/pull/2124

This PR changes how change listeners work. Currently we will trigger all synchronous change listeners on the same thread when doing a `commitTransaction` on that thread.

This is unfortunate when #2124 is implemented as that will pause the autoupdate of RealmResults for the event loop (or until refresh is called) which means that if the change listener was triggered you could potentially access deleted objects when reacting to the change.

Triggering all change listeners from the Looper is also more consistent, but does introduce the slight weirdness that change listeners added after a commit will also be triggered. However, change listeners are inherently asynchronous so the change in this behaviour shouldn't cause regressions. The biggest impact is potentially on unit tests that depended on the exact time a listener was triggered.

### CODE REVIEW

**Comment by cmelchior** [comment]
Ready for review @nhachicha @realm/java 
Note that most changes happened in our type based unit tests that depended on the knowledge that local listeners where triggered immediately.

File: changelog.txt [hunk 1]
```diff
@@ -1,7 +1,8 @@
 0.88.0
  * BREAKING CHANGE: DynamicRealm.executeTransaction() now directly throws any RuntimeException instead of wrapping it in a RealmException (#1682).
  * BREAKING CHANGE: DynamicRealm.executeTransaction() now throws IllegalArgumentException instead of silently accepting a null Transaction object.
- * BREAKING CHANGE: String setters now throws IllegalArgumentException instead of RealmError for invalid surrogates. 
+ * BREAKING CHANGE: String setters now throws IllegalArgumentException instead of RealmError for invalid surrogates.
```
Comment (kneth) [inline changelog.txt]: "throws" -> "throw"

**Comment by cmelchior** [comment]
Reverting to WIP, some unit tests are failing spectacular on CI but not on my device.

**Comment by cmelchior** [comment]
Ready for review @realm/java

File: realm/realm-library/src/main/java/io/realm/HandlerController.java [hunk 2]
```diff
@@ -112,6 +112,7 @@ public boolean handleMessage(Message message) {
                 }
                 case COMPLETED_UPDATE_ASYNC_QUERIES: {
                     // this is called once the background thread completed the update of the async queries
+                    notifyAllListeners();
```
Comment (beeender) [inline realm/realm-library/src/main/java/io/realm/HandlerController.java]: Is this correct? `completedAsyncQueriesUpdate` will emit `REALM_CHANGED` again, and `notifyAllListeners()` will be called anyway.

File: realm/realm-library/src/main/java/io/realm/Realm.java [hunk 3]
```diff
@@ -1093,7 +1093,7 @@ public void run() {
                     transaction.execute(bgRealm);
 
                     if (!Thread.currentThread().isInterrupted()) {
-                        bgRealm.commitTransaction(new Runnable() {
+                        bgRealm.commitTransaction(true, new Runnable() {
```
Comment (beeender) [inline realm/realm-library/src/main/java/io/realm/Realm.java]: `true` here is actually redundant. Here, the realm is supposed to be closed before sending notifications, which means, even true is passed here, the local realm won't be notified because of it is supposed to be closed before.

File: realm/realm-library/src/main/java/io/realm/RealmQuery.java [hunk 4]
```diff
@@ -18,6 +18,7 @@
 
 
 import android.os.Handler;
+import android.util.Log;
```
Comment (beeender) [inline realm/realm-library/src/main/java/io/realm/RealmQuery.java]: Log is not used.

File: realm/realm-library/src/main/java/io/realm/internal/async/RealmThreadPoolExecutor.java [hunk 5]
```diff
@@ -81,6 +81,9 @@ protected void beforeExecute(Thread t, Runnable r) {
         }
     }
 
+    /**
+     * Pause the executor. Pausing means the executor will stop starting new tasks (but complete current ones).
```
Comment (beeender) [inline realm/realm-library/src/main/java/io/realm/internal/async/RealmThreadPoolExecutor.java]: `Pause` -> `Pauses`

File: realm/realm-library/src/androidTest/java/io/realm/NotificationsTest.java [hunk 6]
```diff
@@ -551,7 +582,9 @@ public void onChange() {
         RealmChangeListener listenerB = new RealmChangeListener() {
             @Override
             public void onChange() {
-                counter.incrementAndGet();
+                assertEquals(0, counter.get());
```
Comment (beeender) [inline realm/realm-library/src/androidTest/java/io/realm/NotificationsTest.java]: Now we rely on the order of listener invocations. Worth to create another test case to ensure that ListenerA is called first.

File: realm/realm-library/src/androidTest/java/io/realm/TypeBasedNotificationsTests.java [hunk 7]
```diff
@@ -501,95 +482,108 @@ public void run() {
         dog.addChangeListener(new RealmChangeListener() {
             @Override
             public void onChange() {
-                assertEquals("Akamaru", dog.getName());
+                // Step 4: Respond to relevant change
                 typebasedCommitInvocations.incrementAndGet();
+                assertEquals("Akamaru", dog.getName());
+                assertEquals(17, dog.getAge());
```
Comment (beeender) [inline realm/realm-library/src/androidTest/java/io/realm/TypeBasedNotificationsTests.java]: call `typebasedCommitInvocations.incrementAndGet();` after assertions, to avoid flaky test.

Comment (beeender) [inline realm/realm-library/src/androidTest/java/io/realm/TypeBasedNotificationsTests.java]: Uh, never mind, they are in the same thread.

File: realm/realm-library/src/androidTest/java/io/realm/NotificationsTest.java (hunk 6 above)
Comment (cmelchior) [inline realm/realm-library/src/androidTest/java/io/realm/NotificationsTest.java]: Good idea. Added.

File: realm/realm-library/src/main/java/io/realm/Realm.java (hunk 3 above)
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/Realm.java]: Good catch

File: realm/realm-library/src/main/java/io/realm/HandlerController.java (hunk 2 above)
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/HandlerController.java]: I _think_ it is OK since REALM_CHANGED will only be triggered if the caller thread has advanced since it started its async query, but I hope @nhachicha will review this throughly.

**Comment by cmelchior** [comment]
Feedback fixed @realm/java

**Comment by beeender** [comment]
`commitTransaction` in `BaseRealm.migrateRealm()` doesn't seem to be necessary to trigger notifications. Maybe change that as well?

File: realm/realm-library/src/main/java/io/realm/BaseRealm.java [hunk 8]
```diff
@@ -201,7 +201,8 @@ public void removeAllChangeListeners() {
         handlerController.removeAllChangeListeners();
     }
 
-    void setHandler (Handler handler) {
+    // WARNING: If this method is used after calling a any async method, the old handler will still be used.
```
Comment (nhachicha) [inline realm/realm-library/src/main/java/io/realm/BaseRealm.java]: `s/calling a any/calling any`
also we can highlight that this is for test only

File: realm/realm-library/src/main/java/io/realm/HandlerController.java [hunk 9]
```diff
@@ -331,23 +332,27 @@ private void realmChanged() {
...
+        }
+        if (!realm.isClosed()) {
+            notifyRealmObjectCallbacks();
+        }
+
+        // empty async RealmObject shouldn't block the realm to advance
+        // they're empty so no risk on running into a corrupt state
+        // where the pointer (Row) is using one version of a Realm, whereas the
+        // current Realm is advancing to a newer version (they're empty anyway)
+        if (!realm.isClosed() && threadContainsAsyncEmptyRealmObject()) {
```
Comment (nhachicha) [inline realm/realm-library/src/main/java/io/realm/HandlerController.java]: I would keep this outside of this method, the method's responsibility is to only notify listeners

File: realm/realm-library/src/main/java/io/realm/HandlerController.java (hunk 2 above)
Comment (nhachicha) [inline realm/realm-library/src/main/java/io/realm/HandlerController.java]: I'm not sure about this either.
The current implementation of `completedAsyncQueriesUpdate` will notify the listeners if we advance the Realm.
```java
// notify listeners only when we advanced
            if (compare != 0) {
                notifyGlobalListeners();
                // notify RealmResults & RealmObject callbacks (type based notifications)
                notifySyncRealmResultsCallbacks();
                notifyRealmObjectCallbacks();
            }
```
maybe you can refactor it to 
```java
 if (compare != 0) {
    notifyAllListeners();
  }
```

**Comment by nhachicha** [comment]
@cmelchior I refactored some of the code related to notifications to make it easy to reason about listeners notifications. I also fixed some failing tests.

The idea is to use `notifyAllListeners` when possible this will simply notify `notifyGlobalListeners` & `notifyTypeBasedListeners` this way it's easy to follow which action trigger which notification

File: realm/realm-library/src/androidTest/java/io/realm/NotificationsTest.java [hunk 10]
```diff
@@ -567,11 +601,33 @@ public void onChange() {
...
+
+            @Override
+            public void onChange() {
+                firstListenerCalled.countDown();
+            }
+        });
+        realm.addChangeListener(new RealmChangeListener() {
+            @Override
+            public void onChange() {
+                TestHelper.awaitOrFail(firstListenerCalled);
```
Comment (beeender) [inline realm/realm-library/src/androidTest/java/io/realm/NotificationsTest.java]: They are in the same thread, no need to wait. just assert.

File: realm/realm-library/src/main/java/io/realm/HandlerController.java [hunk 11]
```diff
@@ -469,13 +466,11 @@ private void completedAsyncQueriesUpdate(QueryUpdateTask.Result result) {
...
 
-            // notify listeners only when we advanced
-            if (compare != 0) {
-                notifyGlobalListeners();
-                // notify RealmResults & RealmObject callbacks (type based notifications)
-                notifySyncRealmResultsCallbacks();
-                notifyRealmObjectCallbacks();
-            }
+            // We need to notify the rest of listeners, since the original REALM_CHANGE
+            // was delayed/swallowed in order to be able to update async queries
```
Comment (beeender) [inline realm/realm-library/src/main/java/io/realm/HandlerController.java]: @nhachicha Why did you remove the `if (compare != 0) {`? if `compare == 0`, `advanceRead` is not called, and no changes in this `SharedGroup` instance.

**Comment by cmelchior** [comment]
Updated PR @realm/java

File: realm/realm-library/src/androidTest/java/io/realm/NotificationsTest.java [hunk 12]
```diff
@@ -567,11 +601,32 @@ public void onChange() {
...
+        realm.addChangeListener(new RealmChangeListener() {
+            @Override
+            public void onChange() {
+                assertTrue(firstListenerCalled.get());
+                looperThread.testComplete();
+            }
+        });
+
+        realm.beginTransaction();
+        realm.commitTransaction();
```
Comment (beeender) [inline realm/realm-library/src/androidTest/java/io/realm/NotificationsTest.java]: oops, just realized this is not enough. need to test the other order as well in case the container changes to eg. `Set`.

Comment (cmelchior) [inline realm/realm-library/src/androidTest/java/io/realm/NotificationsTest.java]: Not sure what you mean?

Comment (beeender) [inline realm/realm-library/src/androidTest/java/io/realm/NotificationsTest.java]: It could be accidentally pass the test if we only test `addListener(listener1)` `addListener(listener2)`, to avoid that we have to also test the other way around with same `listener1` and `listener2` : `addListener(listener2)` `addListener(listener1)`

**Comment by cmelchior** [comment]
Updated the unit test @beeender / @realm/java

**Comment by beeender** [comment]
https://github.com/realm/realm-java/pull/2233/files#r53263681 Still have one pending comments? @nhachicha

File: realm/realm-library/src/main/java/io/realm/HandlerController.java (hunk 11 above)
Comment (nhachicha) [inline realm/realm-library/src/main/java/io/realm/HandlerController.java]: Like I mentioned in the comment I've added. This `updateAsyncEmptyRealmObject` is a consequence of a `REALM_CHANGE` we delayed/swallowed, so even if we don't advance the Realm, we still need to update sync RealmResults & RealmObject (because we originally delayed their notif  until this completes)

Comment (beeender) [inline realm/realm-library/src/main/java/io/realm/HandlerController.java]: OK. But what about `notifyGlobalListeners()`, shouldn't it like:
```
if (compare != 0) {
    notifyGlobalListeners();
}
notifySyncRealmResultsCallbacks();
notifyRealmObjectCallbacks();
```

**Comment by beeender** [comment]
I have a flaky test `callback_with_relevant_commit_from_different_looper_realmobject_async` with this change now :disappointed:

**Comment by cmelchior** [comment]
It was already flaky before. I actually convinced myself that this PR fixed it, but guess not :(

**Comment by beeender** [comment]
This still have some flaky tests... I am trying to fix now.

File: realm/realm-library/src/main/java/io/realm/HandlerController.java (hunk 11 above)
Comment (beeender) [inline realm/realm-library/src/main/java/io/realm/HandlerController.java]: :+1:  Just figured out this could happen when the transaction is committed in the same thread which has async queries.

**Comment by beeender** [comment]
:+1:  and @realm/java please take a look at my flaky testing in this PR.

**Comment by cmelchior** [comment]
Changes to unit tests looks good. Just one more :+1: @realm/java

**Comment by dalinaum** [comment]
:+1:

**Comment by cmelchior** [comment]
@dalinaum This was merged before being squashed :/
